File: src/Servers/Chat/src/Aggregate/KeyValueManager.cs
using System.Collections.Concurrent;
using System.Collections.Generic;


namespace UniSpy.Server.Chat.Aggregate
{
    public class KeyValueManager
    {
        public ConcurrentDictionary<string, string> Dict { get; private set; } = new ConcurrentDictionary<string, string>();

        public KeyValueManager()
        {
        }
        
        public void Update(Dictionary<string, string> data)
        {
            // TODO check if all key is send through the request or
            // TODO only updated key send through the request
            foreach (var key in data.Keys)
            {
                //we update the key value
                Dict[key] = data[key];
            }
        }
        public void Update(KeyValuePair<string, string> data)
        {
            Dict[data.Key] = data.Value;
        }
        public static string BuildKeyValueString(Dictionary<string, string> keyValues)
        {
            string flags = "";
            foreach (var kv in keyValues)
            {
                flags += $@"\{kv.Key}\{kv.Value}";
            }
            return flags;
        }
        public string GetValueString(List<string> keys)
        {
            string values = "";
            foreach (var key in keys)
            {
                if (Dict.ContainsKey(key))
                {
                    values += @"\" + Dict[key];
                }
                else
                {
                    values += @"\";
                    // throw new ChatException($"Can not find key: {key}");
                }
            }
            return values;
        }

        public bool IsContainAllKey(List<string> keys)
        {
            foreach (var key in keys)
            {
                if (!Dict.ContainsKey(key))
                {
                    return false;
                }
            }
            return true;
        }
    }
}

File: src/servers/chat/src/aggregates/key_value_manager.py
class KeyValueManager:
    data: dict
    """
    store the key and values
    """

    def __init__(self):
        self.data = {}

    def update(self, data: dict):
        for key, value in data.items():
            self.data[key] = value

    def build_key_value_string(self, key_values: dict):
        flags = ""
        for key, value in key_values.items():
            flags += f"\\{key}\\{value}"
        return flags

    def get_value_string(self, keys: list[str]) -> str:
        values = ""
        for key in keys:
            if key in self.data:
                values += f"\\{self.data[key]}"
            else:
                values += "\\"
                # Uncomment the line below to raise an exception if key is not found
                # raise Exception(f"Can not find key: {key}")
        return values

    def is_contain_all_key(self, keys: list[str]):
        return all(key in self.data for key in keys)
